### PR TITLE
Add storage mining pool to genesis and implement automatic reward redeeming

### PR DESCRIPTION
**Pull request by sagar-solana**
#### Problem

Replicators lose unclaimed rewards after some time

#### Summary of Changes

Add optional mining_pool argument to redeem rewards automatically

### CODE REVIEW

File: client/src/thin_client.rs [hunk 1]
```diff
@@ -306,11 +307,31 @@ impl SyncClient for ThinClient {
...
+    fn get_account(&self, pubkey: &Pubkey) -> TransportResult<Option<Account>> {
+        Ok(self.rpc_client().get_account(pubkey).ok())
+    }
+
     fn get_balance(&self, pubkey: &Pubkey) -> TransportResult<u64> {
         let balance = self.rpc_client().get_balance(pubkey)?;
         Ok(balance)
     }
 
+    fn get_recent_blockhash(&self) -> TransportResult<(Hash, FeeCalculator)> {
```
Comment (sagar-solana) [inline client/src/thin_client.rs]: CLion complained that the trait implementation was not `in order` of the definition, it automatically moved this fn to match the trait definition

File: genesis/src/main.rs [hunk 2]
```diff
@@ -151,6 +151,14 @@ fn main() -> Result<(), Box<dyn error::Error>> {
                 .required(true)
                 .help("Path to file containing the bootstrap leader's storage keypair"),
         )
+        .arg(
+            Arg::with_name("storage_mining_pool_keypair_file")
+                .long("storage-mining-pool-keypair")
+                .value_name("STORAGE MINING POOL KEYPAIR")
```
Comment (mvines) [inline genesis/src/main.rs]: Just KEYPAIR is enough here imo

File: replicator/src/main.rs [hunk 3]
```diff
@@ -68,6 +77,12 @@ fn main() {
     } else {
         Keypair::new()
     };
+    let storage_mining_pool_pubkey =
+        matches
+            .value_of("storage_mining_pool_pubkey")
+            .map(|pubkey_str| {
```
Comment (mvines) [inline replicator/src/main.rs]: Duplicate this helper function into this file for better error handling: https://github.com/solana-labs/solana/blob/master/install/src/lib.rs#L28

File: genesis/src/main.rs [hunk 4]
```diff
@@ -151,6 +151,14 @@ fn main() -> Result<(), Box<dyn error::Error>> {
...
+                .takes_value(true)
+                .required(true)
+                .help("Path to file containing the bootstrap leader's storage keypair"),
```
[the remark this answers is not in the document]
Comment (sagar-solana) [reply genesis/src/main.rs]: Yeah, clearly I got distracted half way down. I'll fix it.

File: genesis/src/main.rs [hunk 5]
```diff
@@ -301,6 +313,10 @@ fn main() -> Result<(), Box<dyn error::Error>> {
                     1,
                 ),
             ),
+            (
+                storage_mining_keypair.pubkey(),
+                storage_contract::create_mining_pool_account(100_000_000),
```
Comment (pgarg66) [inline genesis/src/main.rs]: Should we pass number of lamports as command line argument?

Comment (sagar-solana) [reply genesis/src/main.rs]: Yeah i guess I should.

File: replicator/src/main.rs (hunk 3 above)
Comment (sagar-solana) [reply replicator/src/main.rs]: thank you for this. much better

**Comment by rob-solana** [comment]
awesome

**Pull request merged**